### PR TITLE
build: add cel rule to make resource immutable

### PR DESCRIPTION
**Pull request by parth-gr**
added the cel rules at sunvolumgroup and rados
namesapce spec

closes: #13315

**Which issue is resolved by this Pull Request:**
Resolves # #13315

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by subhamkrai — CHANGES_REQUESTED** [review]
Please update the `message`

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 1]
```diff
@@ -2923,12 +2923,14 @@ type CephFilesystemSubVolumeGroupList struct {
 // CephFilesystemSubVolumeGroupSpec represents the specification of a Ceph Filesystem SubVolumeGroup
 type CephFilesystemSubVolumeGroupSpec struct {
 	// The name of the subvolume group. If not set, the default is the name of the subvolumeGroup CR.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="dataSource is immutable"
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply pkg/apis/ceph.rook.io/v1/types.go]: Updated

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 2]
```diff
@@ -2973,10 +2975,12 @@ type CephBlockPoolRadosNamespaceList struct {
 // CephBlockPoolRadosNamespaceSpec represents the specification of a CephBlockPool Rados Namespace
 type CephBlockPoolRadosNamespaceSpec struct {
 	// The name of the CephBlockPoolRadosNamespaceSpec namespace. If not set, the default is the name of the CR.
+	// +kubebuilder:validation:XValidation:message="CephBlockPoolRadosNamespaceSpec name is immutable",rule="self == oldSelf"
 	// +optional
 	Name string `json:"name,omitempty"`
 	// BlockPoolName is the name of Ceph BlockPool. Typically it's the name of
 	// the CephBlockPool CR.
+	// +kubebuilder:validation:XValidation:message="CephBlockPoolRadosNamespaceSpec BlockPoolName is immutable",rule="self == oldSelf"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2983]: I'd suggest no need to include the type name in the message, and let's use the lower case yaml name
```suggestion
	// +kubebuilder:validation:XValidation:message="blockPoolName is immutable",rule="self == oldSelf"
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 3]
```diff
@@ -2973,10 +2975,12 @@ type CephBlockPoolRadosNamespaceList struct {
 // CephBlockPoolRadosNamespaceSpec represents the specification of a CephBlockPool Rados Namespace
 type CephBlockPoolRadosNamespaceSpec struct {
 	// The name of the CephBlockPoolRadosNamespaceSpec namespace. If not set, the default is the name of the CR.
+	// +kubebuilder:validation:XValidation:message="CephBlockPoolRadosNamespaceSpec name is immutable",rule="self == oldSelf"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2978]:
```suggestion
	// +kubebuilder:validation:XValidation:message="name is immutable",rule="self == oldSelf"
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 4]
```diff
@@ -2923,12 +2923,14 @@ type CephFilesystemSubVolumeGroupList struct {
...
 type CephFilesystemSubVolumeGroupSpec struct {
 	// The name of the subvolume group. If not set, the default is the name of the subvolumeGroup CR.
+	// +kubebuilder:validation:XValidation:message="CephFilesystemSubVolumeGroupSpec name is immutable",rule="self == oldSelf"
 	// +optional
 	Name string `json:"name,omitempty"`
 	// FilesystemName is the name of Ceph Filesystem SubVolumeGroup volume name. Typically it's the name of
 	// the CephFilesystem CR. If not coming from the CephFilesystem CR, it can be retrieved from the
 	// list of Ceph Filesystem volumes with `ceph fs volume ls`. To learn more about Ceph Filesystem
 	// abstractions see https://docs.ceph.com/en/latest/cephfs/fs-volumes/#fs-volumes-and-subvolumes
+	// +kubebuilder:validation:XValidation:message="CephFilesystemSubVolumeGroupSpec FilesystemName is immutable",rule="self == oldSelf"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2933]:
```suggestion
	// +kubebuilder:validation:XValidation:message="filesystemName is immutable",rule="self == oldSelf"
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 5]
```diff
@@ -2923,12 +2923,14 @@ type CephFilesystemSubVolumeGroupList struct {
 // CephFilesystemSubVolumeGroupSpec represents the specification of a Ceph Filesystem SubVolumeGroup
 type CephFilesystemSubVolumeGroupSpec struct {
 	// The name of the subvolume group. If not set, the default is the name of the subvolumeGroup CR.
+	// +kubebuilder:validation:XValidation:message="CephFilesystemSubVolumeGroupSpec name is immutable",rule="self == oldSelf"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2926]:
```suggestion
	// +kubebuilder:validation:XValidation:message="name is immutable",rule="self == oldSelf"
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (parth-gr) [reply pkg/apis/ceph.rook.io/v1/types.go]: updated

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 6]
```diff
@@ -2923,12 +2923,14 @@ type CephFilesystemSubVolumeGroupList struct {
...
 type CephFilesystemSubVolumeGroupSpec struct {
 	// The name of the subvolume group. If not set, the default is the name of the subvolumeGroup CR.
+	// +kubebuilder:validation:XValidation:message="name is immutable",rule="self == oldSelf"
 	// +optional
 	Name string `json:"name,omitempty"`
 	// FilesystemName is the name of Ceph Filesystem SubVolumeGroup volume name. Typically it's the name of
 	// the CephFilesystem CR. If not coming from the CephFilesystem CR, it can be retrieved from the
 	// list of Ceph Filesystem volumes with `ceph fs volume ls`. To learn more about Ceph Filesystem
 	// abstractions see https://docs.ceph.com/en/latest/cephfs/fs-volumes/#fs-volumes-and-subvolumes
+	// +kubebuilder:validation:XValidation:message="FilesystemName is immutable",rule="self == oldSelf"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2933]:
```suggestion
	// +kubebuilder:validation:XValidation:message="filesystemName is immutable",rule="self == oldSelf"
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 7]
```diff
@@ -2973,10 +2975,12 @@ type CephBlockPoolRadosNamespaceList struct {
 // CephBlockPoolRadosNamespaceSpec represents the specification of a CephBlockPool Rados Namespace
 type CephBlockPoolRadosNamespaceSpec struct {
 	// The name of the CephBlockPoolRadosNamespaceSpec namespace. If not set, the default is the name of the CR.
+	// +kubebuilder:validation:XValidation:message="name is immutable",rule="self == oldSelf"
 	// +optional
 	Name string `json:"name,omitempty"`
 	// BlockPoolName is the name of Ceph BlockPool. Typically it's the name of
 	// the CephBlockPool CR.
+	// +kubebuilder:validation:XValidation:message="BlockPoolName is immutable",rule="self == oldSelf"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2983]:
```suggestion
	// +kubebuilder:validation:XValidation:message="blockPoolName is immutable",rule="self == oldSelf"
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Review by travisn — APPROVED** [review]

**Review by subhamkrai — APPROVED** [review]

**Pull request merged**